File: python/poker.py
#!/usr/bin/env python3
"""
Skillers.gg Poker Agent — plays Heads-Up No-Limit Texas Hold'em via LLM.

Uses REST to join a game, then WebSocket for real-time gameplay.

Quick start:
  export SKILLERS_API_KEY=sk_agent_xxx
  export OPENAI_API_KEY=sk-xxx        # or ANTHROPIC_API_KEY or GEMINI_API_KEY
  python3 python/poker.py

Customize your strategy by editing the SYSTEM_PROMPT and build_prompt() function.
"""

import os, sys, json, re, requests
import websockets.sync.client

# ── Configuration ────────────────────────────────────────────────────────────
API_URL  = os.environ.get("SKILLERS_API_URL", "https://skillers.gg/api")
WS_URL   = os.environ.get("SKILLERS_WS_URL", "wss://ws.skillers.gg")
API_KEY  = os.environ.get("SKILLERS_API_KEY", "")
LLM_PROVIDER = os.environ.get("LLM_PROVIDER", "openai")     # "openai", "anthropic", or "gemini"
LLM_MODEL    = os.environ.get("LLM_MODEL", "")               # defaults per provider below

DEFAULT_MODELS = {
    "openai": "gpt-4o",
    "anthropic": "claude-sonnet-4-20250514",
    "gemini": "gemini-2.5-flash",
}

# ── System prompt — tweak this to change your agent's personality ────────────
SYSTEM_PROMPT = """You are an expert poker player competing in Heads-Up No-Limit Texas Hold'em.
You play aggressively with strong hands, bluff occasionally, and make mathematically sound decisions.
Consider hand strength, pot odds, position, and opponent tendencies.
Always respond with ONLY a JSON object — no explanation, no markdown."""

# ── LLM Integration ─────────────────────────────────────────────────────────

def ask_llm(prompt: str) -> str:
    """Send a prompt to the configured LLM and return the response text."""
    model = LLM_MODEL or DEFAULT_MODELS.get(LLM_PROVIDER, "gpt-4o")

    if LLM_PROVIDER == "openai":
        key = os.environ.get("OPENAI_API_KEY", "")
        resp = requests.post("https://api.openai.com/v1/chat/completions",
            headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"},
            json={"model": model, "max_tokens": 256, "temperature": 0.3,
                  "messages": [{"role": "system", "content": SYSTEM_PROMPT},
                               {"role": "user", "content": prompt}]},
            timeout=30)
        return resp.json()["choices"][0]["message"]["content"].strip()

    elif LLM_PROVIDER == "anthropic":
        key = os.environ.get("ANTHROPIC_API_KEY", "")
        resp = requests.post("https://api.anthropic.com/v1/messages",
            headers={"x-api-key": key, "Content-Type": "application/json",
                     "anthropic-version": "2023-06-01"},
            json={"model": model, "max_tokens": 256, "system": SYSTEM_PROMPT,
                  "messages": [{"role": "user", "content": prompt}]},
            timeout=30)
        return resp.json()["content"][0]["text"].strip()

    elif LLM_PROVIDER == "gemini":
        key = os.environ.get("GEMINI_API_KEY", "")
        resp = requests.post(
            f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={key}",
            headers={"Content-Type": "application/json"},
            json={"systemInstruction": {"parts": [{"text": SYSTEM_PROMPT}]},
                  "contents": [{"parts": [{"text": prompt}]}],
                  "generationConfig": {"maxOutputTokens": 256, "temperature": 0.3}},
            timeout=30)
        return resp.json()["candidates"][0]["content"]["parts"][0]["text"].strip()

    else:
        raise ValueError(f"Unknown LLM_PROVIDER: {LLM_PROVIDER}")

def parse_json(text: str):
    """Extract JSON from LLM response, handling markdown code blocks."""
    text = re.sub(r"```(?:json)?\s*\n?([\s\S]*?)```", r"\1", text).strip()
    m = re.search(r"\{[\s\S]*\}", text)
    if m:
        try:
            return json.loads(m.group())
        except json.JSONDecodeError:
            pass
    return None

# ── Game-specific prompt builder ─────────────────────────────────────────────

def build_prompt(state: dict, side: str) -> str:
    """Build the prompt sent to your LLM each turn. Edit this to improve play."""
    my_stack  = state["stackA"] if side == "a" else state["stackB"]
    opp_stack = state["stackB"] if side == "a" else state["stackA"]
    my_bet    = state.get("currentBetA", 0) if side == "a" else state.get("currentBetB", 0)
    opp_bet   = state.get("currentBetB", 0) if side == "a" else state.get("currentBetA", 0)
    to_call   = opp_bet - my_bet
    is_dealer = state.get("dealer") == side

    history = state.get("bettingHistory", [])
    history_str = " ".join(
        f"{h['side']}:{h['action']}" + (f"({h['amount']})" if h.get('amount') else "")
        for h in history[-8:]
    ) if history else "none"

    return f"""Heads-Up No-Limit Hold'em — Hand #{state.get('handNumber', 1)}, Stage: {state.get('stage', '?')}
Position: {"Dealer/SB" if is_dealer else "Big Blind"}
Hole cards: {' '.join(state.get('holeCards', []))}
Community:  {' '.join(state.get('community', [])) or '(none)'}
Pot: {state.get('pot', 0)} | Your stack: {my_stack} | Opponent stack: {opp_stack}
To call: {to_call if to_call > 0 else '0 (no bet to match)'}
Recent actions: {history_str}

Legal actions: {'"fold", "call", "raise"' if to_call > 0 else '"check", "raise"'}
For raise, amount is the TOTAL bet size (minimum raise = current bet + big blind).

Respond with ONLY one JSON object:
{{"action": "fold"}} or {{"action": "check"}} or {{"action": "call"}} or {{"action": "raise", "amount": <number>}}"""

# ── Move decision ────────────────────────────────────────────────────────────

def decide_move(state: dict, side: str) -> dict:
    """Ask the LLM for a poker move. Fallback to check/call on failure."""
    my_bet  = state.get("currentBetA", 0) if side == "a" else state.get("currentBetB", 0)
    opp_bet = state.get("currentBetB", 0) if side == "a" else state.get("currentBetA", 0)
    to_call = opp_bet - my_bet

    try:
        prompt = build_prompt(state, side)
        response = ask_llm(prompt)
        move = parse_json(response)
        if move and "action" in move:
            if move["action"] == "raise" and "amount" in move:
                move["amount"] = max(1, int(move["amount"]))
            return move
    except Exception as e:
        print(f"  LLM error: {e}")

    return {"action": "call"} if to_call > 0 else {"action": "check"}

# ── WebSocket game loop ─────────────────────────────────────────────────────

def play_game(game_id: str):
    """Play a full poker game over WebSocket."""
    url = f"{WS_URL}/parties/game-room-server/{game_id}?api_key={API_KEY}"
    with websockets.sync.client.connect(url, close_timeout=5) as ws:
        side = None
        moves = 0
        last_state = None
        move_pending = False

        while moves < 500:
            try:
                raw = ws.recv(timeout=30)
            except TimeoutError:
                ws.send(json.dumps({"type": "ping"}))
                continue

            msg = json.loads(raw)

            if msg["type"] == "authenticated":
                side = msg.get("side")
                if msg.get("game_id"):
                    print(f"  Playing as side {side}")
                continue

            if msg["type"] == "state_update":
                state = msg.get("state", {})
                last_state = state
                if not side or state.get("toAct") != side or move_pending:
                    continue

                move = decide_move(state, side)
                s = state
                print(f"  Hand {s.get('handNumber','?')} [{s.get('stage','?')}] "
                      f"Cards: {s.get('holeCards',[])} Board: {s.get('community',[])} → {move}")
                ws.send(json.dumps({"type": "move", "move": move}))
                move_pending = True
                moves += 1
                continue

            if msg["type"] == "move_accepted":
                move_pending = False
                if msg.get("gameOver"):
                    winner = msg.get("winner_agent_id", "draw")
                    print(f"\nGame over after {moves} moves! Winner: {winner}")
                    return
                continue

            if msg["type"] == "move_rejected":
                move_pending = False
                error = msg.get("error", "?")
                print(f"  Move rejected: {error}")
                # Only retry for actionable rejections (illegal move, bad format, etc.)
                if any(skip in error.lower() for skip in ["not your turn", "internal", "already being processed"]):
                    continue
                if side and last_state and last_state.get("toAct") == side:
                    el = error.lower()
                    if "cannot check" in el or "bet of" in el or "minimum raise" in el:
                        fallback = {"action": "call"}
                    elif "cannot call" in el:
                        fallback = {"action": "check"}
                    else:
                        fallback = {"action": "fold"}
                    ws.send(json.dumps({"type": "move", "move": fallback}))
                    move_pending = True
                continue

            if msg["type"] == "game_over":
                print(f"\nGame over! Winner: {msg.get('winner_id', '?')}")
                return

            if msg["type"] == "error":
                print(f"  WS error: {msg.get('message', msg)}")
                if msg.get("code") in ("invalid_key", "auth_required"):
                    return
                continue

        print(f"\nReached {moves} moves limit.")

# ── Main ─────────────────────────────────────────────────────────────────────

def main():
    if not API_KEY:
        print("Set SKILLERS_API_KEY env var. Get one at https://skillers.gg/docs")
        sys.exit(1)

    model = LLM_MODEL or DEFAULT_MODELS.get(LLM_PROVIDER, "?")
    print(f"Skillers Poker Agent — LLM: {LLM_PROVIDER}/{model}")
    print(f"Joining poker game...")

    r = requests.post(f"{API_URL}/games/join",
        headers={"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"},
        json={"game_type": "poker", "room_amount_cents": 0})
    r.raise_for_status()
    join = r.json()

    game_id = join["game_id"]
    print(f"Game: {game_id} ({join['status']})")

    if join["status"] == "waiting":
        print("Waiting for opponent... (will be notified via WebSocket)")

    play_game(game_id)

    # Show final result
    game = requests.get(f"{API_URL}/games/{game_id}").json()
    print(f"Result: {game.get('status')} | Winner: {game.get('winner_agent_id', 'none')}")

if __name__ == "__main__":
    main()

File: typescript/poker.ts
#!/usr/bin/env npx tsx
/**
 * Skillers.gg Poker Agent — plays Heads-Up No-Limit Texas Hold'em via LLM.
 *
 * Quick start:
 *   export SKILLERS_API_KEY=sk_agent_xxx
 *   export OPENAI_API_KEY=sk-xxx        # or ANTHROPIC_API_KEY or GEMINI_API_KEY
 *   npx tsx typescript/poker.ts
 *
 * Customize your strategy by editing SYSTEM_PROMPT and buildPrompt().
 */

// ── Configuration ───────────────────────────────────────────────────────────
const API_URL      = process.env.SKILLERS_API_URL || "https://skillers.gg/api";
const API_KEY      = process.env.SKILLERS_API_KEY || "";
const LLM_PROVIDER = process.env.LLM_PROVIDER || "openai";    // "openai", "anthropic", "gemini"
const LLM_MODEL    = process.env.LLM_MODEL || "";
const POLL_MS      = 1000;

const DEFAULT_MODELS: Record<string, string> = {
  openai: "gpt-4o",
  anthropic: "claude-sonnet-4-20250514",
  gemini: "gemini-2.5-flash",
};

// ── System prompt — tweak this to change your agent's personality ───────────
const SYSTEM_PROMPT = `You are an expert poker player competing in Heads-Up No-Limit Texas Hold'em.
You play aggressively with strong hands, bluff occasionally, and make mathematically sound decisions.
Consider hand strength, pot odds, position, and opponent tendencies.
Always respond with ONLY a JSON object — no explanation, no markdown.`;

// ── LLM Integration ────────────────────────────────────────────────────────

async function askLLM(prompt: string): Promise<string> {
  const model = LLM_MODEL || DEFAULT_MODELS[LLM_PROVIDER] || "gpt-4o";

  if (LLM_PROVIDER === "openai") {
    const key = process.env.OPENAI_API_KEY || "";
    const res = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: { Authorization: `Bearer ${key}`, "Content-Type": "application/json" },
      body: JSON.stringify({ model, max_tokens: 256, temperature: 0.3,
        messages: [{ role: "system", content: SYSTEM_PROMPT }, { role: "user", content: prompt }] }),
    });
    const data = await res.json() as any;
    return data.choices[0].message.content.trim();
  }

  if (LLM_PROVIDER === "anthropic") {
    const key = process.env.ANTHROPIC_API_KEY || "";
    const res = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: { "x-api-key": key, "Content-Type": "application/json", "anthropic-version": "2023-06-01" },
      body: JSON.stringify({ model, max_tokens: 256, system: SYSTEM_PROMPT,
        messages: [{ role: "user", content: prompt }] }),
    });
    const data = await res.json() as any;
    return data.content[0].text.trim();
  }

  if (LLM_PROVIDER === "gemini") {
    const key = process.env.GEMINI_API_KEY || "";
    const res = await fetch(
      `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${key}`,
      { method: "POST", headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ systemInstruction: { parts: [{ text: SYSTEM_PROMPT }] },
          contents: [{ parts: [{ text: prompt }] }],
          generationConfig: { maxOutputTokens: 256, temperature: 0.3 } }) });
    const data = await res.json() as any;
    return data.candidates[0].content.parts[0].text.trim();
  }

  throw new Error(`Unknown LLM_PROVIDER: ${LLM_PROVIDER}`);
}

function parseJSON(text: string): any | null {
  text = text.replace(/```(?:json)?\s*\n?([\s\S]*?)```/g, "$1").trim();
  const m = text.match(/\{[\s\S]*\}/);
  if (m) try { return JSON.parse(m[0]); } catch {}
  return null;
}

// ── Skillers API ────────────────────────────────────────────────────────────

const headers = () => ({ Authorization: `Bearer ${API_KEY}`, "Content-Type": "application/json" });

async function apiGet(path: string) {
  const r = await fetch(`${API_URL}${path}`, { headers: headers() });
  return r.json() as any;
}

async function apiPost(path: string, body: any) {
  const r = await fetch(`${API_URL}${path}`, { method: "POST", headers: headers(), body: JSON.stringify(body) });
  if (!r.ok) { const err = await r.json().catch(() => ({})) as any; throw Object.assign(new Error(err.error || r.statusText), { body: err }); }
  return r.json() as any;
}

async function apiPostRaw(path: string, body: any) {
  return fetch(`${API_URL}${path}`, { method: "POST", headers: headers(), body: JSON.stringify(body) });
}

// ── Prompt builder ──────────────────────────────────────────────────────────

function buildPrompt(stateData: any): string {
  const s = stateData.state;
  const side = stateData.your_side;
  const myStack  = side === "a" ? s.stackA : s.stackB;
  const oppStack = side === "a" ? s.stackB : s.stackA;
  const myBet    = side === "a" ? (s.currentBetA || 0) : (s.currentBetB || 0);
  const oppBet   = side === "a" ? (s.currentBetB || 0) : (s.currentBetA || 0);
  const toCall   = oppBet - myBet;
  const isDealer = s.dealer === side;

  const history = (s.bettingHistory || []).slice(-8)
    .map((h: any) => `${h.side}:${h.action}${h.amount ? `(${h.amount})` : ""}`)
    .join(" ") || "none";

  return `Heads-Up No-Limit Hold'em — Hand #${s.handNumber || 1}, Stage: ${s.stage || "?"}
Position: ${isDealer ? "Dealer/SB" : "Big Blind"}
Hole cards: ${(s.holeCards || []).join(" ")}
Community:  ${(s.community || []).join(" ") || "(none)"}
Pot: ${s.pot || 0} | Your stack: ${myStack} | Opponent stack: ${oppStack}
To call: ${toCall > 0 ? toCall : "0 (no bet to match)"}
Recent actions: ${history}

Legal actions: ${toCall > 0 ? '"fold", "call", "raise"' : '"check", "raise"'}
For raise, amount is the TOTAL bet size (minimum raise = current bet + big blind).

Respond with ONLY one JSON object:
{"action": "fold"} or {"action": "check"} or {"action": "call"} or {"action": "raise", "amount": <number>}`;
}

// ── Move decision ───────────────────────────────────────────────────────────

async function decideMove(stateData: any): Promise<any> {
  const s = stateData.state;
  const side = stateData.your_side;
  const myBet  = side === "a" ? (s.currentBetA || 0) : (s.currentBetB || 0);
  const oppBet = side === "a" ? (s.currentBetB || 0) : (s.currentBetA || 0);
  const toCall = oppBet - myBet;

  try {
    const prompt = buildPrompt(stateData);
    const response = await askLLM(prompt);
    const move = parseJSON(response);
    if (move?.action) {
      if (move.action === "raise" && move.amount) move.amount = Math.max(1, Math.floor(move.amount));
      return move;
    }
  } catch (e: any) {
    console.log(`  LLM error: ${e.message}`);
  }

  return toCall > 0 ? { action: "call" } : { action: "check" };
}

// ── Game loop ───────────────────────────────────────────────────────────────

async function playGame(gameId: string) {
  let moves = 0;
  while (moves < 500) {
    await new Promise(r => setTimeout(r, POLL_MS));
    const stateData = await apiGet(`/games/${gameId}/state`);

    if (stateData.status !== "active") {
      console.log(`\nGame ended: ${stateData.status}`);
      return;
    }
    if (!stateData.your_turn) { process.stdout.write("."); continue; }

    const move = await decideMove(stateData);
    const s = stateData.state;
    console.log(`\n  Hand ${s.handNumber || "?"} [${s.stage || "?"}] Cards: ${(s.holeCards||[]).join(" ")} Board: ${(s.community||[]).join(" ")} → ${JSON.stringify(move)}`);

    try {
      const result = await apiPost(`/games/${gameId}/move`, move);
      moves++;
      if (result.gameOver) { console.log(`\nGame over after ${moves} moves! Winner: ${result.winner_agent_id || "draw"}`); return; }
    } catch (e: any) {
      console.log(`\n  Move rejected: ${e.body?.error || e.message}`);
      for (const fallback of [{ action: "call" }, { action: "check" }, { action: "fold" }]) {
        try {
          const result = await apiPost(`/games/${gameId}/move`, fallback);
          moves++;
          if (result.gameOver) return;
          break;
        } catch { continue; }
      }
    }
  }
}

// ── Main ────────────────────────────────────────────────────────────────────

async function main() {
  if (!API_KEY) { console.log("Set SKILLERS_API_KEY env var. Get one at https://skillers.gg/docs"); process.exit(1); }

  const model = LLM_MODEL || DEFAULT_MODELS[LLM_PROVIDER] || "?";
  console.log(`Skillers Poker Agent — LLM: ${LLM_PROVIDER}/${model}`);

  const join = await apiPost("/games/join", { game_type: "poker", room_amount_cents: 0 });
  console.log(`Game: ${join.game_id} (${join.status})`);

  if (join.status === "waiting") {
    console.log("Waiting for opponent...");
    while (true) {
      await new Promise(r => setTimeout(r, 2000));
      const s = await apiGet(`/games/${join.game_id}/state`);
      if (s.status === "active") { console.log("Matched!"); break; }
      if (s.status && s.status !== "waiting") { console.log(`Game cancelled: ${s.status}`); return; }
    }
  }

  await playGame(join.game_id);

  const game = await (await fetch(`${API_URL}/games/${join.game_id}`)).json() as any;
  console.log(`Result: ${game.status} | Winner: ${game.winner_agent_id || "none"}`);
}

main().catch(console.error);
